File: code/packages/elixir/document_ast/test/document_ast_test.exs
defmodule CodingAdventures.DocumentAstTest do
  use ExUnit.Case, async: true
  alias CodingAdventures.DocumentAst

  doctest CodingAdventures.DocumentAst

  describe "block node constructors" do
    test "document/1 creates a document node" do
      assert DocumentAst.document([]) == %{type: :document, children: []}
    end

    test "document/1 with children" do
      para = DocumentAst.paragraph([DocumentAst.text("hello")])
      doc = DocumentAst.document([para])
      assert doc == %{type: :document, children: [%{type: :paragraph, children: [%{type: :text, value: "hello"}]}]}
    end

    test "heading/2 creates heading nodes for all levels" do
      for level <- 1..6 do
        h = DocumentAst.heading(level, [])
        assert h.type == :heading
        assert h.level == level
        assert h.children == []
      end
    end

    test "paragraph/1 creates a paragraph node" do
      p = DocumentAst.paragraph([DocumentAst.text("content")])
      assert p.type == :paragraph
      assert length(p.children) == 1
    end

    test "code_block/2 creates a code block" do
      cb = DocumentAst.code_block("elixir", "IO.puts(1)\n")
      assert cb == %{type: :code_block, language: "elixir", value: "IO.puts(1)\n"}
    end

    test "code_block/2 allows nil language" do
      cb = DocumentAst.code_block(nil, "some code\n")
      assert cb.language == nil
    end

    test "blockquote/1 creates a blockquote" do
      bq = DocumentAst.blockquote([])
      assert bq == %{type: :blockquote, children: []}
    end

    test "list/4 creates an unordered list" do
      lst = DocumentAst.list(false, nil, true, [])
      assert lst == %{type: :list, ordered: false, start: nil, tight: true, children: []}
    end

    test "list/4 creates an ordered list with start" do
      lst = DocumentAst.list(true, 3, false, [])
      assert lst == %{type: :list, ordered: true, start: 3, tight: false, children: []}
    end

    test "list_item/1 creates a list item" do
      item = DocumentAst.list_item([])
      assert item == %{type: :list_item, children: []}
    end

    test "thematic_break/0 creates a thematic break" do
      tb = DocumentAst.thematic_break()
      assert tb == %{type: :thematic_break}
    end

    test "raw_block/2 creates a raw block" do
      rb = DocumentAst.raw_block("html", "<div>raw</div>\n")
      assert rb == %{type: :raw_block, format: "html", value: "<div>raw</div>\n"}
    end
  end

  describe "inline node constructors" do
    test "text/1 creates a text node" do
      t = DocumentAst.text("Hello & world")
      assert t == %{type: :text, value: "Hello & world"}
    end

    test "emphasis/1 creates an emphasis node" do
      em = DocumentAst.emphasis([DocumentAst.text("hello")])
      assert em.type == :emphasis
      assert length(em.children) == 1
    end

    test "strong/1 creates a strong node" do
      s = DocumentAst.strong([DocumentAst.text("bold")])
      assert s.type == :strong
    end

    test "code_span/1 creates a code span" do
      cs = DocumentAst.code_span("const x = 1")
      assert cs == %{type: :code_span, value: "const x = 1"}
    end

    test "link/3 creates a link node" do
      ln = DocumentAst.link("https://example.com", "Example", [DocumentAst.text("click")])
      assert ln.type == :link
      assert ln.destination == "https://example.com"
      assert ln.title == "Example"
    end

    test "link/3 allows nil title" do
      ln = DocumentAst.link("https://example.com", nil, [])
      assert ln.title == nil
    end

    test "image/3 creates an image node" do
      img = DocumentAst.image("cat.png", nil, "a cat")
      assert img == %{type: :image, destination: "cat.png", title: nil, alt: "a cat"}
    end

    test "autolink/2 creates an autolink node" do
      al = DocumentAst.autolink("user@example.com", true)
      assert al.type == :autolink
      assert al.is_email == true
    end

    test "raw_inline/2 creates a raw inline node" do
      ri = DocumentAst.raw_inline("html", "<em>raw</em>")
      assert ri == %{type: :raw_inline, format: "html", value: "<em>raw</em>"}
    end

    test "hard_break/0 creates a hard break" do
      hb = DocumentAst.hard_break()
      assert hb == %{type: :hard_break}
    end

    test "soft_break/0 creates a soft break" do
      sb = DocumentAst.soft_break()
      assert sb == %{type: :soft_break}
    end
  end

  describe "nested document structure" do
    test "can build a complete document tree" do
      doc =
        DocumentAst.document([
          DocumentAst.heading(1, [DocumentAst.text("Title")]),
          DocumentAst.paragraph([
            DocumentAst.text("Hello "),
            DocumentAst.emphasis([DocumentAst.text("world")])
          ]),
          DocumentAst.list(false, nil, true, [
            DocumentAst.list_item([
              DocumentAst.paragraph([DocumentAst.text("item 1")])
            ])
          ])
        ])

      assert doc.type == :document
      assert length(doc.children) == 3
      assert hd(doc.children).type == :heading
    end
  end
end

File: code/packages/dart/document-ast/test/document_ast_test.dart
import 'package:coding_adventures_document_ast/document_ast.dart';
import 'package:test/test.dart';

void main() {
  group('block nodes', () {
    test('builds the core TE00 document structure', () {
      final heading = HeadingNode(1, [TextNode('Document AST')]);
      final paragraph = ParagraphNode([
        TextNode('Hello '),
        EmphasisNode([TextNode('portable')]),
        const SoftBreakNode(),
        StrongNode([TextNode('world')]),
      ]);
      final document = DocumentNode([heading, paragraph]);

      expect(document.type, 'document');
      expect(document.children, hasLength(2));
      expect(document.children.first, same(heading));
      expect(heading.type, 'heading');
      expect(heading.level, 1);
      expect(paragraph.type, 'paragraph');
      expect(paragraph.children.map((node) => node.type), [
        'text',
        'emphasis',
        'soft_break',
        'strong',
      ]);
    });

    test('supports all heading levels and rejects invalid levels', () {
      for (var level = 1; level <= 6; level++) {
        expect(HeadingNode(level, []).level, level);
      }
      expect(() => HeadingNode(0, []), throwsArgumentError);
      expect(() => HeadingNode(7, []), throwsArgumentError);
    });

    test('represents literal, quotation, and raw blocks', () {
      final code = CodeBlockNode(language: 'dart', value: 'final x = 1;\n');
      final unlabelled = CodeBlockNode(language: null, value: 'plain\n');
      final quote = BlockquoteNode([
        ParagraphNode([TextNode('quoted')]),
        BlockquoteNode([]),
      ]);
      final raw = RawBlockNode(format: 'latex', value: r'\textbf{bold}');

      expect(code.type, 'code_block');
      expect(code.language, 'dart');
      expect(code.value, endsWith('\n'));
      expect(unlabelled.language, isNull);
      expect(quote.type, 'blockquote');
      expect(quote.children.last.type, 'blockquote');
      expect(raw.type, 'raw_block');
      expect(raw.format, 'latex');
      expect(raw.value, r'\textbf{bold}');
      expect(const ThematicBreakNode().type, 'thematic_break');
    });

    test('represents regular and task list items', () {
      final regular = ListItemNode([
        ParagraphNode([TextNode('first')]),
      ]);
      final task = TaskItemNode(
        checked: true,
        children: [
          ParagraphNode([TextNode('ship it')])
        ],
      );
      final unordered = ListNode(
        ordered: false,
        start: null,
        tight: true,
        children: [regular, task],
      );
      final ordered = ListNode(
        ordered: true,
        start: 42,
        tight: false,
        children: [regular],
      );

      expect(unordered.type, 'list');
      expect(unordered.children.map((node) => node.type), [
        'list_item',
        'task_item',
      ]);
      expect(unordered.start, isNull);
      expect(unordered.tight, isTrue);
      expect(task.checked, isTrue);
      expect(ordered.start, 42);
      expect(ordered.ordered, isTrue);
    });

    test('represents the stabilized table model', () {
      final header = TableRowNode(
        isHeader: true,
        children: [
          TableCellNode([TextNode('Name')])
        ],
      );
      final body = TableRowNode(
        isHeader: false,
        children: [
          TableCellNode([
            StrongNode([TextNode('Ada')])
          ]),
        ],
      );
      final table = TableNode(
        align: [TableAlignment.left],
        children: [header, body],
      );

      expect(table.type, 'table');
      expect(table.align, [TableAlignment.left]);
      expect(table.children, hasLength(2));
      expect(table.children.first.type, 'table_row');
      expect(table.children.first.isHeader, isTrue);
      expect(table.children.first.children.first.type, 'table_cell');
      expect(table.children.last.children.first.children.first.type, 'strong');
      expect(TableAlignment.values.map((value) => value.name), [
        'left',
        'right',
        'center',
      ]);
    });
  });

  group('inline nodes', () {
    test('represents text and nested formatting', () {
      final text = TextNode('Hello & κόσμε');
      final emphasis = EmphasisNode([text, CodeSpanNode('&amp;')]);
      final strong = StrongNode([emphasis]);
      final strike = StrikethroughNode([strong]);

      expect(text.type, 'text');
      expect(text.value, 'Hello & κόσμε');
      expect(emphasis.type, 'emphasis');
      expect((emphasis.children.last as CodeSpanNode).value, '&amp;');
      expect(strong.type, 'strong');
      expect(strike.type, 'strikethrough');
    });

    test('represents resolved links, images, and autolinks', () {
      final link = LinkNode(
        destination: 'https://example.com',
        title: 'Example',
        children: [TextNode('visit')],
      );
      final untitled = LinkNode(
        destination: '/relative',
        title: null,
        children: [],
      );
      final image = ImageNode(
        destination: 'cat.png',
        title: null,
        alt: 'a tabby cat',
      );
      final email = AutolinkNode(
        destination: 'user@example.com',
        isEmail: true,
      );
      final url = AutolinkNode(
        destination: 'https://example.com',
        isEmail: false,
      );

      expect(link.type, 'link');
      expect(link.title, 'Example');
      expect(link.children.single.type, 'text');
      expect(untitled.title, isNull);
      expect(image.type, 'image');
      expect(image.alt, 'a tabby cat');
      expect(email.type, 'autolink');
      expect(email.isEmail, isTrue);
      expect(url.isEmail, isFalse);
    });

    test('represents raw content and break leaves', () {
      final raw = RawInlineNode(format: 'html', value: '<em>raw</em>');
      const hard = HardBreakNode();
      const soft = SoftBreakNode();

      expect(raw.type, 'raw_inline');
      expect(raw.format, 'html');
      expect(raw.value, '<em>raw</em>');
      expect(hard.type, 'hard_break');
      expect(soft.type, 'soft_break');
    });
  });

  group('sealed discriminators', () {
    test('covers every block-node discriminator', () {
      final nodes = <BlockNode>[
        DocumentNode([]),
        HeadingNode(1, []),
        ParagraphNode([]),
        CodeBlockNode(language: null, value: ''),
        BlockquoteNode([]),
        ListNode(ordered: false, start: null, tight: true, children: []),
        ListItemNode([]),
        TaskItemNode(checked: false, children: []),
        const ThematicBreakNode(),
        RawBlockNode(format: 'html', value: ''),
        TableNode(align: [], children: []),
        TableRowNode(isHeader: false, children: []),
        TableCellNode([]),
      ];

      expect(nodes.map(_classify), [
        'document',
        'heading',
        'paragraph',
        'code_block',
        'blockquote',
        'list',
        'list_item',
        'task_item',
        'thematic_break',
        'raw_block',
        'table',
        'table_row',
        'table_cell',
      ]);
    });

    test('covers every inline-node discriminator', () {
      final nodes = <InlineNode>[
        TextNode(''),
        EmphasisNode([]),
        StrongNode([]),
        StrikethroughNode([]),
        CodeSpanNode(''),
        LinkNode(destination: '/', title: null, children: []),
        ImageNode(destination: '/', title: null, alt: ''),
        AutolinkNode(destination: 'x@y.example', isEmail: true),
        RawInlineNode(format: 'html', value: ''),
        const HardBreakNode(),
        const SoftBreakNode(),
      ];

      expect(nodes.map(_classify), [
        'text',
        'emphasis',
        'strong',
        'strikethrough',
        'code_span',
        'link',
        'image',
        'autolink',
        'raw_inline',
        'hard_break',
        'soft_break',
      ]);
    });
  });

  group('immutability and value semantics', () {
    test('defensively copies child and alignment lists', () {
      final inlineChildren = <InlineNode>[TextNode('before')];
      final paragraph = ParagraphNode(inlineChildren);
      inlineChildren.add(TextNode('after'));

      final blockChildren = <BlockNode>[paragraph];
      final document = DocumentNode(blockChildren);
      blockChildren.clear();

      final align = <TableAlignment?>[TableAlignment.center, null];
      final table = TableNode(align: align, children: []);
      align[0] = TableAlignment.left;

      expect(paragraph.children, hasLength(1));
      expect(document.children, [paragraph]);
      expect(table.align, [TableAlignment.center, null]);
      expect(
          () => paragraph.children.add(TextNode('no')), throwsUnsupportedError);
      expect(() => document.children.clear(), throwsUnsupportedError);
      expect(
          () => table.align.add(TableAlignment.right), throwsUnsupportedError);
    });

    test('uses structural equality and stable hash codes', () {
      final left = DocumentNode([
        HeadingNode(2, [TextNode('Title')]),
        ParagraphNode([TextNode('Body'), const SoftBreakNode()]),
      ]);
      final right = DocumentNode([
        HeadingNode(2, [TextNode('Title')]),
        ParagraphNode([TextNode('Body'), const SoftBreakNode()]),
      ]);
      final different = DocumentNode([
        HeadingNode(3, [TextNode('Title')]),
      ]);

      expect(left, equals(right));
      expect(left.hashCode, right.hashCode);
      expect(left, isNot(equals(different)));
      expect(TextNode('same'), equals(TextNode('same')));
      expect(TextNode('same'), isNot(equals(TextNode('different'))));
    });

    test('applies value semantics to every stabilized node variant', () {
      final pairs = <(Node, Node)>[
        (DocumentNode([]), DocumentNode([])),
        (HeadingNode(2, []), HeadingNode(2, [])),
        (ParagraphNode([]), ParagraphNode([])),
        (
          const CodeBlockNode(language: 'dart', value: 'x\n'),
          const CodeBlockNode(language: 'dart', value: 'x\n'),
        ),
        (BlockquoteNode([]), BlockquoteNode([])),
        (
          ListNode(
            ordered: true,
            start: 3,
            tight: false,
            children: [],
          ),
          ListNode(
            ordered: true,
            start: 3,
            tight: false,
            children: [],
          ),
        ),
        (ListItemNode([]), ListItemNode([])),
        (
          TaskItemNode(checked: true, children: []),
          TaskItemNode(checked: true, children: []),
        ),
        (const ThematicBreakNode(), const ThematicBreakNode()),
        (
          const RawBlockNode(format: 'html', value: '<hr>'),
          const RawBlockNode(format: 'html', value: '<hr>'),
        ),
        (
          TableNode(align: [TableAlignment.right, null], children: []),
          TableNode(align: [TableAlignment.right, null], children: []),
        ),
        (
          TableRowNode(isHeader: true, children: []),
          TableRowNode(isHeader: true, children: []),
        ),
        (TableCellNode([]), TableCellNode([])),
        (const TextNode('text'), const TextNode('text')),
        (EmphasisNode([]), EmphasisNode([])),
        (StrongNode([]), StrongNode([])),
        (StrikethroughNode([]), StrikethroughNode([])),
        (const CodeSpanNode('code'), const CodeSpanNode('code')),
        (
          LinkNode(destination: '/', title: 'home', children: []),
          LinkNode(destination: '/', title: 'home', children: []),
        ),
        (
          const ImageNode(destination: 'x.png', title: null, alt: 'x'),
          const ImageNode(destination: 'x.png', title: null, alt: 'x'),
        ),
        (
          const AutolinkNode(destination: 'x@y.test', isEmail: true),
          const AutolinkNode(destination: 'x@y.test', isEmail: true),
        ),
        (
          const RawInlineNode(format: 'html', value: '<b>x</b>'),
          const RawInlineNode(format: 'html', value: '<b>x</b>'),
        ),
        (const HardBreakNode(), const HardBreakNode()),
        (const SoftBreakNode(), const SoftBreakNode()),
      ];

      expect(pairs.map((pair) => pair.$1.type).toSet(), hasLength(24));
      for (final (left, right) in pairs) {
        expect(left, equals(right), reason: left.type);
        expect(left.hashCode, right.hashCode, reason: left.type);
      }
    });
  });

  test('walks a nested document through typed containment', () {
    final document = DocumentNode([
      HeadingNode(1, [TextNode('Hello')]),
      BlockquoteNode([
        ListNode(
          ordered: false,
          start: null,
          tight: true,
          children: [
            ListItemNode([
              ParagraphNode([
                TextNode('World'),
                const SoftBreakNode(),
                TextNode('!'),
              ]),
            ]),
          ],
        ),
      ]),
    ]);

    final texts = <String>[];
    _collectText(document, texts);
    expect(texts, ['Hello', 'World', '!']);
  });
}

String _classify(Node node) => switch (node) {
      DocumentNode() => 'document',
      HeadingNode() => 'heading',
      ParagraphNode() => 'paragraph',
      CodeBlockNode() => 'code_block',
      BlockquoteNode() => 'blockquote',
      ListNode() => 'list',
      ListItemNode() => 'list_item',
      TaskItemNode() => 'task_item',
      ThematicBreakNode() => 'thematic_break',
      RawBlockNode() => 'raw_block',
      TableNode() => 'table',
      TableRowNode() => 'table_row',
      TableCellNode() => 'table_cell',
      TextNode() => 'text',
      EmphasisNode() => 'emphasis',
      StrongNode() => 'strong',
      StrikethroughNode() => 'strikethrough',
      CodeSpanNode() => 'code_span',
      LinkNode() => 'link',
      ImageNode() => 'image',
      AutolinkNode() => 'autolink',
      RawInlineNode() => 'raw_inline',
      HardBreakNode() => 'hard_break',
      SoftBreakNode() => 'soft_break',
    };

void _collectText(Node node, List<String> values) {
  switch (node) {
    case TextNode(:final value):
      values.add(value);
    case DocumentNode(:final children) ||
          BlockquoteNode(:final children) ||
          ListItemNode(:final children) ||
          TaskItemNode(:final children):
      for (final child in children) {
        _collectText(child, values);
      }
    case HeadingNode(:final children) ||
          ParagraphNode(:final children) ||
          EmphasisNode(:final children) ||
          StrongNode(:final children) ||
          StrikethroughNode(:final children) ||
          LinkNode(:final children) ||
          TableCellNode(:final children):
      for (final child in children) {
        _collectText(child, values);
      }
    case ListNode(:final children):
      for (final child in children) {
        _collectText(child, values);
      }
    case TableNode(:final children):
      for (final child in children) {
        _collectText(child, values);
      }
    case TableRowNode(:final children):
      for (final child in children) {
        _collectText(child, values);
      }
    case CodeBlockNode() ||
          ThematicBreakNode() ||
          RawBlockNode() ||
          CodeSpanNode() ||
          ImageNode() ||
          AutolinkNode() ||
          RawInlineNode() ||
          HardBreakNode() ||
          SoftBreakNode():
      return;
  }
}
